### DISCUSSION TITLE
Feedback: New stateful test runner in CLI

### DISCUSSION THREAD

**Discussion by Stranger6667** [schemathesis/schemathesis#2262] [Q&A]
We've introduced an experimental feature for using the new test runner for stateful testing in CLI. It aims to enhance the quality of stateful testing available in CLI and make the generated scenarios much more diverse.

### How to Enable

CLI option: `--experimental=stateful-test-runner`

### Feedback

We're eager to hear your feedback:
- Did this change help in finding bugs in your APIs?
- Does the output contain enough information to assess the effectiveness of stateful testing?
- Have you encountered any issues or unexpected behavior?
- Any additional feedback or suggestions?

Your feedback is crucial for the improvement of CLI & stateful tests. Thank you for helping us enhance Schemathesis!

**Comment by pregress** [comment]
When using Links with a multi line request body schemathesis give a parse error. 


## Breaks
```yaml

responses:
  '200':
    description: OK
    content:
      application/json:
        schema:
          type: array
          items:
            $ref: '#/components/schemas/fooBar'
    links:
      fooingTheBar:
        operationId: fooBar
        description: example
        requestBody: >
          { 
            "Foo": "{$response.body#/0/HostName}", 
            "Hello": "World" 
          }
```

## Works
```yaml

responses:
  '200':
    description: OK
    content:
      application/json:
        schema:
          type: array
          items:
            $ref: '#/components/schemas/fooBar'
    links:
      fooingTheBar:
        operationId: fooBar
        description: example
        requestBody: {  "Foo": "{$response.body#/0/HostName}",  "Hello": "World" }

```

## Output
```
________________________________ Stateful tests ________________________________
Runtime Error

Nested embedded expressions are not allowed

    Traceback (most recent call last):
      File "/usr/local/lib/python3.12/site-packages/schemathesis/stateful/runner.py", line 261, in _execute_state_machine_loop
        InstrumentedStateMachine.run(settings=config.hypothesis_settings)
      File "/usr/local/lib/python3.12/site-packages/schemathesis/stateful/state_machine.py", line 115, in run
        return run_state_machine_as_test(cls, settings=settings)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/usr/local/lib/python3.12/site-packages/schemathesis/stateful/__init__.py", line 147, in run_state_machine_as_test
        return _run_state_machine_as_test(state_machine_factory, settings=settings, _min_steps=2)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/usr/local/lib/python3.12/site-packages/hypothesis/stateful.py", line 247, in run_state_machine_as_test
        run_state_machine(state_machine_factory)
      File "/usr/local/lib/python3.12/site-packages/hypothesis/stateful.py", line 118, in run_state_machine
        @given(st.data())
                   ^^^^^^^
      File "/usr/local/lib/python3.12/site-packages/hypothesis/core.py", line 1722, in wrapped_test
        raise the_error_hypothesis_found
      File "/usr/local/lib/python3.12/site-packages/schemathesis/specs/openapi/stateful/__init__.py", line 175, in step_function
        return APIStateMachine._step(*args_, **kwargs_)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/usr/local/lib/python3.12/site-packages/schemathesis/stateful/state_machine.py", line 150, in _step
        return self.step(case, (previous, link))
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/usr/local/lib/python3.12/site-packages/schemathesis/stateful/runner.py", line 172, in step
        result = super().step(case, previous)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/usr/local/lib/python3.12/site-packages/schemathesis/stateful/state_machine.py", line 167, in step
        case = self.transform(result, direction, case)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/usr/local/lib/python3.12/site-packages/schemathesis/specs/openapi/stateful/__init__.py", line 38, in transform
        direction.set_data(case, elapsed=result.elapsed, context=context)
      File "/usr/local/lib/python3.12/site-packages/schemathesis/specs/openapi/links.py", line 220, in set_data
        self.set_body(case, context, overrides)
      File "/usr/local/lib/python3.12/site-packages/schemathesis/specs/openapi/links.py", line 269, in set_body
        evaluated = expressions.evaluate(self.body, context, evaluate_nested=True)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/usr/local/lib/python3.12/site-packages/schemathesis/specs/openapi/expressions/__init__.py", line 29, in evaluate
        parts = [node.evaluate(context) for node in parser.parse(expr)]
                                                    ^^^^^^^^^^^^^^^^^^
      File "/usr/local/lib/python3.12/site-packages/schemathesis/specs/openapi/expressions/parser.py", line 14, in parse
        return list(_parse(expr))
               ^^^^^^^^^^^^^^^^^^
      File "/usr/local/lib/python3.12/site-packages/schemathesis/specs/openapi/expressions/parser.py", line 27, in _parse
        raise RuntimeExpressionError("Nested embedded expressions are not allowed")
    schemathesis.specs.openapi.expressions.errors.RuntimeExpressionError: Nested embedded expressions are not allowed
    Falsifying example:
    state = _InstrumentedStateMachine()
    state.teardown()
```

**Reply by Stranger6667** [reply]
The issue is that in YAML, the following is a string:

```yaml
        requestBody: >
          { 
            "Foo": "{$response.body#/0/HostName}", 
            "Hello": "World" 
          }
```

because `>` indicates a multiline string.

```yaml
        requestBody: {  "Foo": "{$response.body#/0/HostName}",  "Hello": "World" }
```

This is parsed as an object and handled correctly. Schemathesis works with Open API expressions here OR with compound types (array & object) it recursively evaluates them. If it is a string, then double `{` will indicate a nested expression which is forbidden

**Comment by pregress** [comment]
The junit output for the stateful testrunner just contains one entry for stateful tests.
Would be nice if it could contain the tested paths, similar to the console output.

```xml
<testsuites disabled="0" errors="0" failures="0" tests="12" time="60.57535773499694">
	<testsuite disabled="0" errors="0" failures="0" name="schemathesis" skipped="0" tests="12" time="60.57535773499694" hostname="367f021c11ac">
                <!--other tests-->
		<testcase name="Stateful tests" time="10.945231"/>
	</testsuite>
</testsuites>

```

**Reply by Stranger6667** [reply]
Thank you so much for reporting! I’ll check it soon

**Reply by Stranger6667** [reply]
I like the idea! However, I'd like to add it there after iterating on the output structure

**Reply by benjmountain** [reply]
I know this is a bit of an old thread, just wanted to see if this feature request might be something added to the new 4.X of Schemathesis? Having those stateful paths listed out in the JUnit report would be a big W when cataloguing test cases

**Reply by Stranger6667** [reply]
Yep, one of the next minor releases will expand stateful testing quite a bit, I'll include this one too as a separate issue

**Comment by BronzeDeer** [comment]
Reusing and shrinking `--hypothesis-phases=reuse,shrink` an existing hypothesis cache with the stateful runner multiple times seems to lead to increasing runtimes and an ever growing ExceptionGroup `Hypothesis found 857 distinct failures. (857 sub-exceptions)`, that schemathesis considers a RuntimeError. Is this just expected over successive runs logging many failures or should hypothesis start tuning its corpus with the output of previous shrink runs? If we disable generating new cases, I'd expect hypothesis to find the minimal example for every error case and on subsequent runs finish very quickly because it only tries the minimal examples and then quits since they are not minimizable anymore 

Unfortunately this behaviour occurs for me in a complicated microservice project with a complicated openapi schema.
I'm not familiar enough with deeper hypothesis and schemathesis workings to isolate this. Any pointers on where to look so I can try and find a minimal reproducible example?

![image](https://github.com/user-attachments/assets/abb1e886-6b5a-47a6-8d60-f2f9fff7f425)

**Reply by Stranger6667** [reply]
> Is this just expected over successive runs logging many failures or should hypothesis start tuning its corpus with the output of previous shrink runs?

My assumption is that there is some failure-deduplication issue, but it should not produce `RuntimeError` in any event. 

Generally, I think your expectations are reasonable, and right now it seems like the approach in this stateful runner does not handle some cases.

> Any pointers on where to look so I can try and find a minimal reproducible example?

I don't have ideas right now, but I'd try to run existing tests in the Schemathesis test suite with those phases. Chances are that this issue occurs within [existing test scenarios](https://github.com/schemathesis/schemathesis/blob/master/test/specs/openapi/stateful/test_runner.py) when they would have this extra reuse & shrink step.

**Comment by pregress** [comment]
When you have the stateful output with links. It would be nice if there was a better way to debug/investigate why some endpoints aren't hit or only return a specific status code.
Personally I was thinking it would be better to concat the `X-Schemathesis-TestCaseId` from the parent to the child. Eg: `PUk6n7-83IzHA`
This way you can trace back which input was used for a child and this would help building better links. 

Or have a seperate flag to see the actual call tree in the output?

Currently:
```
GET /foo/bar/{fooId}
└── 200
    └── GET /foo/bar/{fooId}/{barId}      0     15      0       15
```

Proposal:
```
GET /foo/bar/1
└── 200
    └── GET /foo/bar/1/2               0     1     0       1
    └── GET /foo/bar/3/4               0     1     0       1
```

Maybe gather some more feedback on this?

I think the testcaseId is a better solution as this will also work for request/response body.

**Reply by Stranger6667** [reply]
> When you have the stateful output with links. It would be nice if there was a better way to debug/investigate why some endpoints aren't hit or only return a specific status code.

I completely agree that right now the output does not help with debugging much.

> Personally I was thinking it would be better to concat the X-Schemathesis-TestCaseId from the parent to the child. Eg: PUk6n7-83IzHA

As an extension to your idea, I believe, that the right approach would be to represent cases as a tree (and the whole stateful run as a forest) so the dependencies between cases are clear. There is already a way to refer parents via `source` attribute, but it is inconvenient. 

This way, it will be easy to build a sequence of ids as you mentioned. There are other implications:
- Much cleaner event system in the runner
- More cases to detect `use_after_free` (right now it looks only at ancestors to detect whether DELETE was called)
- Easier to visualize

On the visualization side, I'd like to implement an HTML report where that tree could be represented as a full list as you proposed, or some aggregation. 

> Maybe gather some more feedback on this?

Console output is extremely limiting, especially for large state machines and many generated cases. I could imagine the detailed output would be overwhelming. 

My inclination is to keep the console output small (or even shrink it) and provide a more detailed version in HTML. There it will be easier to have filters to select only specific endpoints, etc.

also, request/response could be hidden in HTML and displayed on demand.

Right now I am finishing some immediate things with the coverage phase and then will focus on the data model for the runner. Meanwhile, I'll prepare a questionnaire and will try to gather feedback on different parts of the output, including your proposal (+ on the future HTML report structure).

When the feedback form is ready I'll create a separate issue for it.

Thank you again for participating :) I appreciate it a lot! Feel free to share any concerns or questions :)

**Comment by adnweedon** [comment]
I've been playing with the new stateful runner, and noticed something interesting. I have a POST endpoint to create a model, linked to a GET and a DELETE endpoint by the model ID - all working as expected. I've just added a new PUT endpoint to the spec to edit the model, linked in the same way, but before implementing it I ran tests, which all passed!

The PUT path includes the model ID, and so 404 is a valid response, because the ID might not exist. When testing statefully, though, 404 shouldn't be a valid response, because we know the model exists!

Have I missed something in how I've set up my spec, or is this a bug with the test runner?

**Reply by Stranger6667** [reply]
@adnweedon 

Sorry, I've missed your question. There were quite a lot of changes to the stateful runner, and it handles transitions much more carefully now. Could you, please, retry running these tests with the latest version to check if the issue is still there?
